cmd/kube-apiserver/app/server_test.go

```diff
@@ -0,0 +1,340 @@
+	}
+
+	l, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		panic(err)
+	}
+	defer l.Close()
+	return l.Addr().(*net.TCPAddr).Port
+}
+
+func tryToStartServer(t *testing.T) (*restclient.Config, chan<- struct{}) {
+	stopCh := make(chan struct{})
+	defer func() {
+		if stopCh != nil {
+			close(stopCh)
+		}
+	}()
+
+	s := options.NewServerRunOptions()
+	s.SecureServing.BindPort = 0
+	s.InsecureServing.BindPort = freePort()
+
+	t.Logf("Starting etcd...")
+	etcdServer, storageConfig := etcdtesting.NewUnsecuredEtcd3TestClientServer(t, api.Scheme)
+	go func() {
+		<-stopCh
+		etcdServer.Terminate(t)
+	}()
+
+	s.Etcd.StorageConfig = *storageConfig
+	s.Etcd.DefaultStorageMediaType = "application/json"
+	s.ServiceClusterIPRange.IP = net.IPv4(10, 0, 0, 0)
+	s.ServiceClusterIPRange.Mask = net.CIDRMask(16, 32)
+
+	t.Logf("Starting kube-apiserver...")
+	runErrCh := make(chan error, 1)
+
+	go func() {
+		err := Run(s, stopCh)
+		if err != nil {
+			t.Logf("kube-apiserver exited uncleanly: %v", err)
+			runErrCh <- err
+		}
+	}()
+
+	t.Logf("Creating client...")
+	config := &restclient.Config{
+		Host: net.JoinHostPort("127.0.0.1", fmt.Sprintf("%d", s.InsecureServing.BindPort)),
+	}
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		t.Fatalf("Failed to create a client: %v", err)
+	}
+
+	t.Logf("Waiting for /healthz to be ok...")
+	err = wait.Poll(100*time.Millisecond, 30*time.Second, func() (bool, error) {
+		select {
+		case err := <-runErrCh:
+			return false, err
+		default:
+		}
+
+		result := client.CoreV1Client.RESTClient().Get().AbsPath("/healthz").Do()
+		status := 0
+		result.StatusCode(&status)
+		if status == 200 {
+			return true, nil
+		}
+		return false, nil
+	})
+	if err != nil && strings.Contains(err.Error(), "bind") {
+		t.Logf("Bind error, retrying...")
+		return nil, nil
+	}
+	if err != nil {
+		t.Fatalf("Failed to wait for /healthz to return ok: %v", err)
+	}
+
+	// transfer stopCh ownership to caller
+	resultCh := stopCh
+	stopCh = nil
+	return config, resultCh
+}
+
+func startServerOrDie(t *testing.T) (*restclient.Config, chan<- struct{}) {
+	// retry test because the bind might fail due to a race with another process
+	// binding to the port. We cannot listen to :0 (then the kernel would give us
+	// a port which is free for sure), so we need this workaround.
+	for retry := 0; retry < 5 && !t.Failed(); retry++ {
+		config, stopCh := tryToStartServer(t)
+		if stopCh != nil {
+			return config, stopCh
+		}
+	}
+
+	t.Fatalf("Failed to launch server")
+	return nil, nil
+}
+
+func TestRun(t *testing.T) {
+	config, stopCh := startServerOrDie(t)
+	defer close(stopCh)
+
+	client, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	// test whether the server is really healthy after /healthz told us so
+	t.Logf("Creating Deployment directly after healthy")
+	var replicas int32 = 1
+	_, err = client.AppsV1beta1().Deployments("default").Create(&v1beta1.Deployment{
```

Comment:
doesn't seem to make a difference after some test runs. How reliable can you reproduce that with kubeadm? Every time?